node/pkg/watchers/near/watcher.go

```diff
@@ -1,556 +1,309 @@
-
-				standard := event_json.Get("standard")
-				if !standard.Exists() || standard.String() != "wormhole" {
-					continue
-				}
-				event_type := event_json.Get("event")
-				if !event_type.Exists() || event_type.String() != "publish" {
-					continue
-				}
-
-				em := event_json.Get("emitter")
-				if !em.Exists() {
-					continue
-				}
-
-				emitter, err := hex.DecodeString(em.String())
-				if err != nil {
-					return err
-				}
-
-				var a vaa.Address
-				copy(a[:], emitter)
-
-				id, err := base58.Decode(hash)
-				if err != nil {
-					return err
-				}
-
-				var txHash = eth_common.BytesToHash(id) // 32 bytes = d3b136a6a182a40554b2fafbc8d12a7a22737c10c81e33b33d1dcb74c532708b
-
-				v := event_json.Get("data")
-				if !v.Exists() {
-					logger.Info("data")
-					return nil
-				}
-
-				pl, err := hex.DecodeString(v.String())
-				if err != nil {
-					return err
-				}
-
-				block_hash_str := block_hash.String()
-
-				txBlock, err := e.getBlockHash(block_hash_str)
-				if err != nil {
-					return err
-				}
-				body := gjson.ParseBytes(txBlock)
-				if !body.Exists() {
-					return errors.New("block parse error")
-				}
-				ts_nanosec := body.Get("result.header.timestamp")
-				if !ts_nanosec.Exists() {
-					return errors.New("block parse error, missing timestamp")
-				}
-				ts := uint64(ts_nanosec.Uint()) / 1000000000
-
-				if e.mainnet {
-					height := body.Get("result.header.height")
-					if height.Exists() && height.Uint() < 74473147 {
-						return errors.New("test missing observe")
-					}
-				}
-
-				observation := &common.MessagePublication{
-					TxHash:           txHash,
-					Timestamp:        time.Unix(int64(ts), 0),
-					Nonce:            uint32(event_json.Get("nonce").Uint()), // uint32
-					Sequence:         event_json.Get("seq").Uint(),
-					EmitterChain:     vaa.ChainIDNear,
-					EmitterAddress:   a,
-					Payload:          pl,
-					ConsistencyLevel: 0,
+			for i := 0; i < len(newJobs); i++ {
+				if e.transactionProcessingQueue.Len() > quequeSize {
+					logger.Error(
+						"NEAR transactionProcessingQueue exceeds max queue size. Skipping transaction.",
+						zap.String("log_msg_type", "tx_proc_queue_full"),
+						zap.String("chunk_id", chunkHeader.Hash),
+					)
+					p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDNear, 1)
+					break
 				}
-
-				nearMessagesConfirmed.Inc()
-
-				logger.Info("message observed",
-					zap.Uint64("ts", ts),
-					zap.Time("timestamp", observation.Timestamp),
-					zap.Uint32("nonce", observation.Nonce),
-					zap.Uint64("sequence", observation.Sequence),
-					zap.Stringer("emitter_chain", observation.EmitterChain),
-					zap.Stringer("emitter_address", observation.EmitterAddress),
-					zap.Binary("payload", observation.Payload),
-					zap.Uint8("consistency_level", observation.ConsistencyLevel),
-				)
-
-				e.msgChan <- observation
+				e.transactionProcessingQueue.Schedule(newJobs[i], time.Now().Add(newJobs[i].delay))
```

Comment:
so, 99.999% of the time.. the transaction will not have a wormhole log message in it at all.  In my version, I just retrieve the txn initially to see if we have any interest in it at all.  Is there really a need to wait 3 seconds to do this?

Reply:
Good idea, but I think we need to wait because at the start of the transaction we don't know which receipts it may generate in the future, and those receipts could include wormhole messages.